tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -307,50 +315,62 @@ std::vector<TensorView*> mlp(
     local_matmul1 = reshape(local_matmul1, {D, B * S, E}, {D, D, B * S / D, E});
   }
   TensorView* matmul1 = sum(local_matmul1, {0}); // Allreduce or Reduce scatter
-  TensorView* linear1 = add(matmul1, broadcast(b1, {true, true, false}));
+  std::vector<bool> bcast_mask(matmul1->nDims() - 1, true);
+  bcast_mask[matmul1->nDims() - 2] = false;
+  TensorView* linear1 = add(matmul1, broadcast(b1, bcast_mask));
   // Dropout
   Val* prob = IrBuilder::create<Val>(1.0 - kDropoutProb);
   Val* scale = IrBuilder::create<Val>(1.0 / (1.0 - kDropoutProb));
-  if (mask == nullptr) {
-    auto rand_vals = rand_like(linear1);
-    mask = lt(rand_vals, prob);
-  }
-  auto apply_mask = mul(linear1, mask);
-  auto dropout_result = mul(apply_mask, scale);
+  TensorView* dropout_result = dropout(linear1, prob, scale).output;
 
   // Tensor parallel shardings
-  for (auto* tv : {w0, b0, w1, linear0, gelu}) {
+  for (auto* tv : {w0, b0, w1}) {
     tv->setDeviceMesh(mesh);
     tv->axis(0)->parallelize(ParallelType::DIDx);
   }
-  for (auto* tv : {x, b1, linear1, dropout_result}) {
+  for (auto* tv : {x, b1}) {
     tv->setDeviceMesh(mesh);
   }
 
   // Sequence parallel shardings
   if (sequence_parallel) {
-    for (auto* tv : {linear1, dropout_result}) {
-      tv->axis(0)->parallelize(ParallelType::DIDx);
-    }
     matmul1->setDeviceMesh(mesh);
     matmul1->axis(1)->parallelize(ParallelType::DIDx);
-    mask->setDeviceMesh(mesh);
-    mask->axis(0)->parallelize(ParallelType::DIDx);
   }
 
-  return {linear0, gelu, linear1, dropout_result};
+  return {linear0, gelu, matmul1, linear1, dropout_result};
```

Comment:
I'm unsure about saving both `matmul1` and `linear1`. Is that a common practice, e.g., done by TE?

Reply:
(I wasn't sure because linear1 could be cheaply rematerialized from matmul1 and b1 -- almost the same compute but much less memory)

Reply:
The fusion does not output matmul1, it's returned by the helper function for sharding propagation, since matmul1 is a boundary tv for reduce-scatter in sequence parallel.